src/data_structures/graph.ml

```diff
@@ -1,61 +1,118 @@
+             Option.value (M.find_opt ind parents_map) ~default:[]
+           in
+
+           let b =
+             match parents with
+             | [] -> List.exists (fun n -> ind = n) entry_points
+             | _ -> true
+           in
+
+           let children, parents, info =
+             if b then (children, parents, info) else ([], [], None)
+           in
+           { ind; info = (info, b); children; parents } )
+         l' )
   in
   { nodes; entry_points }
 
-let rec print_graph nodes (acc, visited) (n : node) =
-  if S.mem n.ind visited then (acc, visited)
-  else
-    let visited = S.add n.ind visited in
-    List.fold_left
-      (fun (acc, visited) (x : int) ->
-        if x < Array.length nodes then
-          let i = Array.get nodes x in
-          print_graph nodes ((n.ind, i.ind) :: acc, visited) i
-        else (acc, visited) )
-      (acc, visited) n.children
+let pp_label fmt s = Fmt.pf fmt {|[label="%a"]|} Fmt.string s
 
-let pp_entry_points fmt l =
-  Fmt.list ~sep:(fun fmt () -> Fmt.pf fmt ";") Fmt.int fmt l
+let pp_label_opt fmt s_opt = Fmt.option pp_label fmt s_opt
 
-let pp_edge fmt (n1, n2) = Fmt.pf fmt "%a -> %a" Fmt.int n1 Fmt.int n2
+let pp_edge n1 fmt (n2, s) =
+  Fmt.pf fmt "%a -> %a %a;\n " Fmt.int n1 Fmt.int n2 pp_label_opt s
 
-let pp_edges fmt l =
-  Fmt.list ~sep:(fun fmt () -> Fmt.pf fmt ";\n") pp_edge fmt l
+let pp_edges fmt (n, l) =
+  Fmt.list ~sep:(fun fmt () -> Fmt.pf fmt "") (pp_edge n) fmt l
 
-let pp_dot fmt g =
-  let entry_points =
-    List.concat_map
-      (fun x ->
-        Option.to_list
-          (if x < Array.length g.nodes then Some (Array.get g.nodes x) else None) )
-      g.entry_points
+let pp_node_cg fmt (n : 'a node) =
+  if snd n.info then
+    Fmt.pf fmt "%a ;\n %a" Fmt.int n.ind pp_edges (n.ind, n.children)
+
+let pp_nodes_cg fmt n =
+  Fmt.array ~sep:(fun fmt () -> Fmt.pf fmt "") pp_node_cg fmt n
+
+let pp_cg_graph fmt g = Fmt.pf fmt "%a" pp_nodes_cg g.nodes
+
+let pp_cg fmt g = Fmt.pf fmt "digraph call_graph {\n %a}" pp_cg_graph g
+
+let init_cfg nodes edges =
+  let children_map, parents_map =
+    List.fold_left
+      (fun (m_children, m_parents) (n1, n2, s) ->
+        (M.add_to_list n1 (n2, s) m_children, M.add_to_list n2 n1 m_parents) )
+      (M.empty, M.empty) edges
+  in
+  let l =
+    List.rev
+      (List.map
+         (fun (ind, info) ->
+           { ind
+           ; info
+           ; children = Option.value (M.find_opt ind children_map) ~default:[]
+           ; parents = Option.value (M.find_opt ind parents_map) ~default:[]
+           } )
+         nodes )
   in
-  let l, _ = List.fold_left (print_graph g.nodes) ([], S.empty) entry_points in
-  Fmt.pf fmt "digraph call_graph {\n%a;\n%a}" pp_entry_points g.entry_points
-    pp_edges l
+  { nodes = Array.of_list l; entry_points = [] }
+
+let pp_inst fmt i = Fmt.pf fmt "%a" (Types.pp_instr ~short:true) i.Annotated.raw
+
+let pp_exp fmt l = Fmt.list ~sep:(fun fmt () -> Fmt.pf fmt " | ") pp_inst fmt l
```

Comment:
```suggestion
let pp_exp fmt l = Fmt.list ~sep:(fun fmt () -> Fmt.string fmt " | ") pp_inst fmt l
```

Can be used when you don't use any of `%a`, `%d` etc. in your formatting string and it should be more efficient (this is nitpicking, feel free to ignore)